### ISSUE TITLE
Question on Calibration

### ISSUE THREAD

**Issue by AllardJM** [microsoft/interpret#11]
I'm testing the classifier algorithm on a dataset (unfortunately confidential) with a binary target (70,000 rows and about 40 predictors) and seeing that while the rank ordering is competitive with other tree based methods, the predictions seem poorly calibrated - even on the training data itself. The prediction is always lower than the actual. **I am wondering if there might be a cause based on the algorithm that could be tuned or if this has been seen in development?**

The model is trained using the default settings (I have tweaked multiple parameters and not found any impact)

```
ebm = ExplainableBoostingClassifier(n_estimators=16,interactions=0,n_jobs=10)
ebm.fit(X,y)
```
The prediction is made on the training data. 
```
p=ebm.predict_proba(X)[:,1]
print(np.mean(p))  # THIS IS 0.023
```
I rank the predictions into deciles (10% bins) and plot the actual target rate and the mean prediction probability for each decile. The rank order is good, AUC is high (this is the training data of course) but we underpredict systematically. The red horizontal line is the overall mean of the training data which is significantly higher than the mean prediction noted above (0.1 versus 0.02)

![image](https://user-images.githubusercontent.com/7504441/57897274-0fc06900-7822-11e9-822d-ebf91b6143ca.png)

**Comment by interpret-ml** [comment]
Hi.  Sorry for the slow reply.  

It looks like you have 16 features?  If so, is the blue curve in the graph that you are comparing to the GAM shape plot the marginalization plot for just that one feature?  Marginalization is when you ignore all other features, and just plot empirical risk as a function of a single feature.  If it is, then we would expect the GAM shape plot to have a smaller scale than marginalization of that feature because the GAM model is careful not to overcount evidence.  In fact this is what makes the GAM model (the full additive model containing shape functions for all 16 features) be well calibrated.  If there was no correlation among the features, then you would expect the marginalization and GAM shape functions to be the same, but usually there is correlation among features and the individual shape plots will (correctly) have smaller range than the corresponding marginalization plots.  

Here's one test you can do: remove the other 15 features and train the GAM on just this one feature.  In that case there is no correlation among features (because there is now only one feature) so there is no opportunity for overcounting evidence.  In this case the GAM shape plot and the marginalization plot should be very similar.

Another way to think of this is suppose we had 10 identical copies of a feature.  Because the features are identical, they would each have the same marginalization plot (i.e., same empirical risk plot).  But if you added all 10 of them together they would now overweight the evidence from this feature by a factor of 10.  The GAM model we train would yield nearly identical shape plots for each of the 10 copies, but each one would receive one tenth of the weight, i.e., have 1/10th of the vertical scale so that when added together they are well calibrated.

Also, although it is common for marginalization plots and GAM shape functions to have somewhat similar shapes (but with different vertical scales), sometimes marginalization and shape plots will differ dramatically.  This is correct behavior, because the shape function for feature i is learned in concert with the shape functions for all other features.  It's as if the shape function for feature i is learned on the residual after all other feature shape functions are trained.  Because of this, it's possible for there to be significant differences between the shape plots and marginalization plots.  This is good --- it means the GAM is learning something more interesting than just marginalization.  If this wasn't the case, we wouldn't need GAMs and instead would just create marginalization plots for each feature and then learn what weights to apply to them.  Sometimes this is actually a pretty good model, but often GAMs learn something more accurate and more interesting.

Hope this helps.  Let us know if you run the experiment suggested above with one feature and don't get the kind of result we're expecting!

Thanks.  -Rich Caruana.

**Comment by BrianMiner** [comment]
Hey Rich-

I am sorry, I think I did a terrible job explaining. 

There are ~40 predictors or so. What is plotted is not any single feature (e.g. pdp plot), instead the plot was created by the following:

- Train the model on X and Y

- Predict the *same* X (probability that y ==1)

- Column stack the true y and the prediction. Call this matrix Z

- Rank Z descending by the prediction column, into 10 equal count bins (deciles). Decile 1 is the highest 10% of the predictions, Decile 10 is the lowest

- Plot the true rate (mean(y)) for each bin (blue line) and the mean prediction (orange line) for each bin.

This is a standard gains or lift type plot. We want to see that the model discriminates (blue is monotonically decreasing from decile 1 to 10) and is well calibrated (orange line is close to the blue at each decile).

For example, training catboost or lightgbm or xgboost etc all have this property - especially on the training set itself. Note: I saw this pattern of under prediction on hold out folds and wanted to confirm that it even occurred with the training set. 

The mean of y in the training set is around 0.1 (10%) while the mean of the predictions is only 0.02 (2%).

There isn't much documentation for this alpha release, so I was wondering if there is a tuning adjustment , if there was something with the prediction (proba) function (maybe the intercept was omitted?) or if this is a known issue with the classifier (in the same way that for example SVMs are often uncalibrated and need Platts Scaling applied to the output).

**Comment by AllardJM** [comment]
Hey Rich-

I am sorry, I think I did a terrible job explaining. 

There are ~40 predictors or so. What is plotted is not any single feature (e.g. pdp plot), instead the plot was created by the following:

- Train the model on X and Y

- Predict the *same* X (probability that y ==1)

- Column stack the true y and the prediction. Call this matrix Z

- Rank Z descending by the prediction column, into 10 equal count bins (deciles). Decile 1 is the highest 10% of the predictions, Decile 10 is the lowest

- Plot the true rate (mean(y)) for each bin (blue line) and the mean prediction (orange line) for each bin.

This is a standard gains or lift type plot. We want to see that the model discriminates (blue is monotonically decreasing from decile 1 to 10) and is well calibrated (orange line is close to the blue at each decile).

For example, training catboost or lightgbm or xgboost etc all have this property - especially on the training set itself. Note: I saw this pattern of under prediction on hold out folds and wanted to confirm that it even occurred with the training set. 

The mean of y in the training set is around 0.1 (10%) while the mean of the predictions is only 0.02 (2%).

There isn't much documentation for this alpha release, so I was wondering if there is a tuning adjustment , if there was something with the prediction (proba) function (maybe the intercept was omitted?) or if this is a known issue with the classifier (in the same way that for example SVMs are often uncalibrated and need Platts Scaling applied to the output).

**Comment by interpret-ml** [comment]
Hi AllardJM, 

No need to apologize -- I misinterpreted (pun intended) your question. Thanks again for all the information. Usually EBM is well calibrated. We just looked at a number of datasets, and found one that's not as well calibrated, which might help us understand the root cause. 

We know you can't send us your data. If we have some questions in the future, would you mind calculating a few statistics for us?

Thanks,
-Rich & the InterpretML Team

**Comment by AllardJM** [comment]
I would be very happy to provide any information I can. It might be possible to share some data as well, would need to gain approval for that.

**Comment by AllardJM** [comment]
I thought perhaps it was an issue of needing to standardize the data, but that did not seem to matter - which is consistent with most if not all tree based models. I then thought I would consider only a handful of features. When picking just a single feature and training the model and again predicting the probability on that same training data set, going through the ranking and plotting described above - the same problem:

![image](https://user-images.githubusercontent.com/7504441/58059796-73a5a300-7b3c-11e9-8d70-900373f17f3c.png)

**Comment by AllardJM** [comment]
I did find that the date_n_epsiodes parameter can control the degree of calibrate (not sure what this does nor if this effect is exclusive to this parameter). With a single feature, the calibration is highly sensitive to the value, even predicting in-sample.

Setting this to 50 over predicts!

```
ExplainableBoostingClassifier(data_n_episodes=50,
               early_stopping_run_length=50,
               early_stopping_tolerance=1e-05, feature_names=['feature_0'],
               feature_step_n_inner_bags=0, feature_types=['continuous'],
               holdout_size=0.15, holdout_split=0.15, interactions=0,
               learning_rate=0.01, max_tree_splits=2,
               min_cases_for_splits=2, n_estimators=16, n_jobs=10,
               random_state=42, schema=None, scoring=None,
               training_step_episodes=1)
```

![image](https://user-images.githubusercontent.com/7504441/58060452-866da700-7b3f-11e9-8f09-86522d878be1.png)

Setting it to 150 underpredicts again....

```
ExplainableBoostingClassifier(data_n_episodes=150,
               early_stopping_run_length=50,
               early_stopping_tolerance=1e-05, feature_names=['feature_0'],
               feature_step_n_inner_bags=0, feature_types=['continuous'],
               holdout_size=0.15, holdout_split=0.15, interactions=0,
               learning_rate=0.01, max_tree_splits=2,
               min_cases_for_splits=2, n_estimators=16, n_jobs=10,
               random_state=42, schema=None, scoring=None,
               training_step_episodes=1)
```
![image](https://user-images.githubusercontent.com/7504441/58060566-ef551f00-7b3f-11e9-9ff7-6e939f70b890.png)


100 gets closer to the right marginal proportion of positive cases (the mean of the predictions is close to the overall rate in the data.) 

```
ExplainableBoostingClassifier(data_n_episodes=100,
               early_stopping_run_length=50,
               early_stopping_tolerance=1e-05, feature_names=['feature_0'],
               feature_step_n_inner_bags=0, feature_types=['continuous'],
               holdout_size=0.15, holdout_split=0.15, interactions=0,
               learning_rate=0.01, max_tree_splits=2,
               min_cases_for_splits=2, n_estimators=16, n_jobs=10,
               random_state=42, schema=None, scoring=None,
               training_step_episodes=1)
```

**Comment by interpret-ml** [comment]
Hi AllardJM, 

Thank you for your detailed notes -- this is really helpful! We are actively experimenting to look into this further. 

Is it possible for you to share the single feature (and corresponding labels) you are using for your experiments? No need to include the feature name, or any other aspects of your dataset -- we'd like to reproduce your results and dive into this one feature problem deeply. 

Thank you again for all your efforts here -- we really appreciate it!

**Comment by AllardJM** [comment]
Yes - if you can provide an email address I can provide some information.

**Comment by interpret-ml** [comment]
Hi @AllardJM, 

Good news -- our latest version of the package (v0.1.8) should have a fix for calibration! We ended up debugging this with another dataset, and found a bug which we believe caused the mis-calibration. If you're interested, you can see our fixes in these commits: https://github.com/microsoft/interpret/commit/b52c8753397129043dd022f39a50de00bb9e810a, https://github.com/microsoft/interpret/commit/ef59b1774fbbfe422f5f011e498f986f8fc9cff1 .

Can you check to see if the latest version produces better results on your dataset? Thank you very much for your hard work in helping us debug this!

-InterpretML Team

**Comment by AllardJM** [comment]
I was getting approval when I saw this....glad you had a source. I pip installed and got similar results as before. There is not a __version__ attributes on interpret so I wasn't sure if I had a new version or not. Is pip updated?

**Comment by interpret-ml** [comment]
Hi @AllardJM - no worries, thanks for getting back to us!

Run `pip show interpret` and see what version it shows. If the version doesn't display there we've run into something pretty unexpected.

Not sure how your pip environment is setup (mirrors, cache etc.). Try `pip install -U interpret --no-cache-dir`.

Currently we're on [v0.1.9](https://pypi.org/project/interpret/) on PyPi. So pip should be updated.

**Comment by AllardJM** [comment]
I had an error installing this time, but added `conda install gxx_linux-64` if anyone else has an issue. Now that I am on the latest version i see much improved results. Still a bit less than say lightgbm but no free lunch etc. Overall, much improved - nice work!
